localboot: added grub and path mode, and registered localbooter for VPD

insomniacslk opened the pull request:
Improvements to local booting.
* Added -grub, -guid, -kernel, -initramfs, -cmdline options to select the boot method
* added LocalBooter and registered in the available booters
* updated uinit to use -grub as default (previously it was an implicit default)

insomniacslk committed:
localboot: added grub and path mode, and registered localbooter for VPD

insomniacslk committed:
Fixed go vet errors

tfg13 commented on localboot/main.go:
```diff
@@ -2,56 +2,43 @@ package main
+// The first parameter, `mountedDevices` is a list of storage.Mountpoint
+// representing a mounted storage device.
+// The second parameter, `dryrun`, will not boot the found configurations if set
+// to true.
```
nit: these seem to have changed names a bit :)

insomniacslk replied:
thanks, fixed in the next commit :)

tfg13 commented on localboot/main.go:
```diff
@@ -2,56 +2,43 @@ package main
+// representing a mounted storage device.
+// The second parameter, `dryrun`, will not boot the found configurations if set
+// to true.
+func BootGrubMode(devices []storage.BlockDev, baseMountpoint string, dryrun bool) error {
```
baseMountpoint hides the outer var, not a huge issue, but maybe consider renaming it

insomniacslk replied:
good point, renamed all the flags to be `flag<NAME>`

tfg13 commented on localboot/main.go:
```diff
@@ -96,10 +74,9 @@ func main() {
-			// note: in dry-run, this loop brea at the first entry
-			// unconditionally
+		debug("Trying boot configuration %+v", cfg)
+		if dryrun {
```
I find it confusing to go into the loop, and then break out immediately. I would put this if outside and never even go in in dryrun mode.

insomniacslk replied:
done

tfg13 commented on localboot/main.go:
```diff
@@ -110,13 +87,116 @@ func main() {
+// look for the given GUID. The second parameter, `guid`, is the partition GUID
+// to look for. The third, `dryrun`, will not actually boot the found
+// configuration if set to true.
+func BootPathMode(devices []storage.BlockDev, baseMountpoint string, guid string, dryrun bool) error {
```
See above, params have changed names, and baseMountpoint hides the outer var again

tfg13 commented on localboot/main.go:
```diff
@@ -110,13 +87,116 @@ func main() {
+	} else {
+		if err := cfg.Boot(); err != nil {
+			cfg.Close()
+			return fmt.Errorf("Failed to boot kernel %s: %v", cfg.Kernel.Name(), err)
```
Is there a chance we ever get here? Do you want to close the files from above before returning in this case? (kernelFD and initramfsFD)

insomniacslk replied:
yes, the boot config may fail and I want to close file descriptors that are not going to be used anymore

tfg13 commented on pkg/booter/localbooter.go:
```diff
@@ -0,0 +1,102 @@
+		`type` is always set to "localboot"
+		`method` can be either "grub" or "path".
+		    The "grub" method will look for grub.cfg or grub2.cfg on the specified device.
+		    If no device is specified, it will look on all the attached storage devices,
```
This sounds like you can use grub mode and still specify which device the grub is installed on. (but the code above always scans all devices instead) This feature would actually be very handy! For some big filesystems, it can even be very expensive to mount them all (double-digit _minutes_). In many cases, people using grub will know where its configuration is, so I think we should support this.

insomniacslk replied:
That's an inconsistency, I will fix it, thanks for spotting!

insomniacslk replied:
and will add a `guid` parameter to the GRUB mode too

tfg13 reviewed (commented):
One feature request/comment, the rest is nits.

insomniacslk committed:
Addressed most of the feedback

insomniacslk commented:
Thanks @tfg13 . I've addressed all of your feedback except one thing in this first commit, and I will add the GUID flag in GRUB mode in the next commit

insomniacslk committed:
Added GUID flag to GRUB mode

tfg13 commented on localboot/main.go:
```diff
@@ -93,30 +119,127 @@ func main() {
+// The second parameter, `baseMountPoint`, is the directory where the mount
+// points for each device will be created.
+// The third parameter, `guid`, is the partition GUID to look for.
+// The third parameter, `dryrun`, will not boot the found configurations if set
```
*fourth

insomniacslk committed:
typo
